src/rdd/rdd.rs

```diff
@@ -568,6 +568,67 @@ pub trait Rdd: RddBase + 'static {
         Ok(buf)
     }
 
+    /// Randomly splits this RDD with the provided weights.
+    fn random_split(
+        &self,
+        weights: Vec<f64>,
+        _seed: Option<u64>,
+    ) -> Vec<SerArc<dyn Rdd<Item = Self::Item>>>
+    where
+        Self: Sized,
+    {
+        let sum: f64 = weights.iter().sum();
+        assert!(
+            weights.iter().all(|&x| x >= 0.0),
+            format!("Weights must be nonnegative, but got {:?}", weights)
+        );
+        assert!(
+            sum > 0.0,
+            format!("Sum of weights must be positive, but got {:?}", weights)
+        );
+
+        let seed: u64 = match _seed {
+            Some(x) => x,
+            None => rand::random::<u64>(),
+        };
+
+
+        let mut full_bounds = vec![0.0f64];
+        let bounds: Vec<f64> =
+            weights
+                .into_iter()
+                .map(|weight| weight / sum)
+                .collect::<Vec<f64>>()
+                .iter()
+                .scan(0.0f64, |state, &x| {
+                    *state = *state + x;
+                    Some(*state)
+                })
+                .collect();
+        full_bounds.extend(bounds);
```

Comment:
I think you can feed the iterator directly to the collection without requiring to collect first.

https://doc.rust-lang.org/stable/std/iter/trait.Extend.html#tymethod.extend

Reply:
Again got compile error after removed the collect(), I need some time to understand it error messages though. For now I will keep it but welcome to refactor it and I would love to learn some iterator tricks.